metrics.cpp

```diff
@@ -0,0 +1,139 @@
+/* This is an example how to add additional evaluation metrics to GraphLab Create using the SDK.
+In this example, we developed seven evaluation metrics function for binary classifier. 
+These functions get as input two SArrays with actual and predicted classifier values, and the label of the positive class, 
+and are able to calculate the classifier's: true-positive (TP), False-Positive (FP),True-Negative (TN),false-negative (FN), precision score, recall score, and f1 score.
+
+The file can be complied by using the following command:
+g++ -std=c++11 metrics.cpp -I <graphlab-sdk directory> -shared -fPIC -o metrics.so
+
+After compilation the file can be loaded to Python using the following line:
+>>> import metrics
+
+Python Code Example:
+>>> import graphlab
+>>> import metrics
+
+
+>>> actual = graphlab.SArray([1,1,0,0,0])
+>>> predicted = graphlab.SArray([0,1,0,0,1])
+>>> metrics.true_positive(actual,predicted, 1) #1
+>>> metrics.false_positive(actual,predicted, 1) # 1
+>>> metrics.true_negative(actual,predicted, 1) # 2
+>>> metrics.false_negative(actual,predicted, 1) # 2
+
+>>> actual = graphlab.SArray(['a','b','a','a','a','b'])
+>>> predicted = graphlab.SArray(['a','a','b','b','b','a'])
+>>> metrics.true_positive(actual,predicted, 'a') # 1
+>>> metrics.false_positive(actual,predicted, 'a') # 2
+>>> metrics.true_negative(actual,predicted, 'a') # 0
+>>> metrics.false_negative(actual,predicted, 'a') # 3
+
+
+For more details please read:
+https://dato.com/products/create/sdk/docs/index.html
+https://github.com/dato-code/GraphLab-Create-SDK
+
+
+Written by: Michael Fire
+*/
+
+#include <graphlab/sdk/toolkit_function_macros.hpp>
+#include <graphlab/sdk/gl_sarray.hpp>
+using namespace graphlab;
+int true_positive(gl_sarray actual, gl_sarray predicted, flexible_type pos_label) {		
+	if(actual.size() != predicted.size())
+	{
+		throw "Error: Cannot calculate True-Positive. Input arrays of different size";
+	}	
+	int tp = 0;
+	for(int i=0;i<actual.size(); i++)
+	{
+		if(actual[i] == pos_label && actual[i] == predicted[i])
+		{
+			tp += 1;
+		}
+		
+	}
+	return tp;
+}
+
+int false_positive(gl_sarray actual, gl_sarray predicted, flexible_type pos_label) {		
```

Comment:
I am far of being an expert in C++ but why using direct access to a array
element is very slow? (can you please give me an example on how to make it
faster?)
In order to make the python version work we also need to add the compiled
CPP version and this make this how-to messier...
it will take me minutes to add this, however, we need to decide if it the
advantages of this addition are bigger than its disadvantage



On Thu, Mar 5, 2015 at 8:57 PM, Srikrishna Sridhar <notifications@github.com
> wrote:

> In metrics.cpp
> <https://github.com/dato-code/how-to/pull/29#discussion_r25890602>:
>
> > +	{
> > +		throw "Error: Cannot calculate True-Positive. Input arrays of different size";
> > +	}	
> > +	int tp = 0;
> > +	for(int i=0;i<actual.size(); i++)
> > +	{
> > +		if(actual[i] == pos_label && actual[i] == predicted[i])
> > +		{
> > +			tp += 1;
> > +		}
> > +		
> > +	}
> > +	return tp;
> > +}
> > +
> > +int false_positive(gl_sarray actual, gl_sarray predicted, flexible_type pos_label) {		
>
> In addition to the above, having a pure python version is much more
> portable than having a C++ SDK version because I can run this code anywhere.
>
> Another option to consider is to add your metrics to the C++ SDK examples
> while adding a pure python evaluation in the how-to?
>
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dato-code/how-to/pull/29/files#r25890602>.
>